Implement `StoreAddressCard` component for contact info

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Partial of #863 and based on #917

- Add a hook to get store address via the core options API
- Add new appearance to **AccountCard** for store address
- Implement the shared **StoreAddressCard** component
- For the **Button** component, fix icon size, and gap between icon and text when icon on right side

💡 Please note that this PR doesn't include the API integration.

### Screenshots:

#### 🎥 . Display store address and refresh

https://user-images.githubusercontent.com/17420811/127481139-34d68d9a-723c-429b-ae75-093183aa853f.mp4

#### 📷 . The preview mode of StoreAddressCard will be used in the Setting page later

![image](https://user-images.githubusercontent.com/17420811/127481897-76d71521-f61b-4d27-8e32-3ae787e8df18.png)

### Detailed test instructions:

1. Go to the MC setup flow and proceed to the "Confirm store requirements" step
2. The store address card should be able to interact
3. Change store address from WC setting page and back to opened MC setup page, the address should be up-to-date after clicking on the refresh icon

### Changelog entry

eason9487 committed:
Fix icon size and gap between icon and text when icon on right side for Button

eason9487 committed:
Add a hook to get store address via the core options API

eason9487 committed:
Add new appearance to AccountCard for store address

eason9487 committed:
Implement the shared StoreAddressCard component

eason9487 committed:
Add StoreAddressCard to ContactInformation

eason9487 commented on js/src/hooks/useStoreAddress.js:
```diff
@@ -0,0 +1,103 @@
+				.every( Boolean );
+
+			let country = '';
+			let state = '';
+
+			if ( loaded ) {
+				const { getSetting } = select( SETTINGS_STORE_NAME );
+				const { countries } = getSetting( 'wc_admin', 'dataEndpoints' );
+				const [ countryCode, stateCode ] = countryAndState.split( ':' );
+
+				const countryMeta = countries.find(
+					( el ) => el.code === countryCode
+				);
+				if ( countryMeta ) {
+					country = countryMeta.name;
+
+					const stateMeta = countryMeta.states.find(
+						( el ) => el.code === stateCode
+					);
+					if ( stateMeta ) {
+						state = stateMeta.name;
+					}
+				}
+			}
+
+			return {
+				refetch,
+				loaded,
+				data: {
+					address,
+					address2,
+					city,
+					state,
+					country,
+					postcode,
+				},
+			};
+		},
+		[ refetch ]
+	);
```
This data source will be replaced with the API implemented in #919. Please ignore it when code review.

ecgan commented on js/src/components/account-card/index.js:
```diff
@@ -5,6 +5,7 @@ import { __ } from '@wordpress/i18n';
 import classnames from 'classnames';
 import { Flex, FlexItem, FlexBlock } from '@wordpress/components';
 import GridiconPhone from 'gridicons/dist/phone';
+import { Icon, store as storeIcon } from '@wordpress/icons';
```
👍  TIL! 

❔ Do you know if there is a published documentation on all the icons, how do they look like? I don't see it in https://www.npmjs.com/package/@wordpress/icons. I'm expecting a page similar to http://automattic.github.io/gridicons/. Also, does this mean that going forward we should use `@wordpress/icons` more, and rely less on `gridicons`?

eason9487 replied:
https://wordpress.github.io/gutenberg/?path=/docs/icons-icon--default
Add this to your bookmark 😎

ecgan replied:
Yes, that's what I'm talking about! Thank you! (They should've mentioned that in the readme. I'll see if I can make that happen.) 😄

ecgan commented on js/src/components/app-button/index.js:
```diff
@@ -49,9 +49,14 @@ const AppButton = ( props ) => {
 
+	const classes = [ 'app-button', className ];
+	if ( rest.iconPosition === 'right' ) {
+		classes.push( 'gla-app-button--fix-icon-position' );
```
❔ Two things here:

1. I think we don't need `gla-app-button`, we can just use back `app-button`.
2. Instead of `fix-icon-position`, I'm thinking a naming like `icon-position-right` would be better.

So to sum it up, I would use `app-button__icon-position-right`.

eason9487 replied:
The `app-button` naming is too general for me and I thought we already had a consensus to have `gla-` prefix for our shared components. About the `fix-icon-position`, I think the **fix** still needs to point out it's a style fix. So I renamed it to `app-button--fix-icon-position-right` by https://github.com/woocommerce/google-listings-and-ads/pull/918/commits/051e2c7f2339b791136607a40ee83b508b17f52e. The `gla-` prefix could be discussed later if you don't think it's needed.

ecgan replied:
> The `app-button` naming is too general for me and I thought we already had a consensus to have `gla-` prefix for our shared components.

The way I think of them is: 

- `app-` prefix: a generic prefix meant for a generic component, it can possibly be used in other projects (copy and paste), or might even be contributed upstream to WordPress or WooCommerce components library.
- `gla-` prefix: very specific to GLA, unlikely to be reused outside of GLA.

Yeah I agree with you that we can discuss this later.

ecgan commented on js/src/components/app-button/index.js:
```diff
@@ -49,9 +49,14 @@ const AppButton = ( props ) => {
 	return (
 		<Button
-			className={ classnames( 'app-button', className ) }
+			className={ classnames( ...classes ) }
```
💅  `classnames` [API](https://www.npmjs.com/package/classnames) is pretty flexible and lenient, you can have it all in one line like the following: (note that I'm putting the two "app-button" class together.)

```suggestion
			className={ classnames(
				'app-button',
				{
					'app-button__icon-position-right':
						rest.iconPosition === 'right',
				},
				className
			) }
```

eason9487 replied:
Thanks! I knew it but I was trying to keep the code easier to read. From my perspective, glue too much logic within JSX does not look better to read.

ecgan replied:
My perspective is that with this suggested change, we won't need to have [Line 52-55](https://github.com/woocommerce/google-listings-and-ads/pull/918/files#diff-ec24036234748771009e50c0fe89c4d4ba61fb4af1484a5162b342e2f1a46640R52-R55) (declaring `classes` array variable and conditionally pushing element into it) which IMO is taking a lil bit more mental energy to read and maintain.

Anyway, this is personal preference 💅 , not a deal breaker, OK to proceed with what we have. 👌

ecgan commented on js/src/components/app-button/index.scss:
```diff
@@ -12,3 +10,14 @@
 }
+
+.gla-app-button {
+	/**
+	 * Fix that the SVG icon gap to text should be on the left side.
+	 * @see https://github.com/WordPress/gutenberg/blob/%40wordpress/components%4012.0.8/packages/components/src/button/style.scss#L309-L311
+	 */
```
👍  Thanks for the code comment. Looking at this comment and the latest https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/button#iconposition (right now it is version 15.0.0), I guess we need this for our current component version 12.0.8 and don't need this for future version.

eason9487 replied:
Not sure your meaning. The [`iconPosition` prop is supported by 12.0.8](https://github.com/WordPress/gutenberg/tree/%40wordpress/components%4012.0.8/packages/components/src/button#iconposition) as well. And that style problem still exists in the latest revision.
- 12.0.8 https://github.com/WordPress/gutenberg/blob/%40wordpress/components%4012.0.8/packages/components/src/button/style.scss#L309-L311
- The latest
https://github.com/WordPress/gutenberg/blob/0e9c8b13611209cd57ae6ed4bcc7d5503d0530a7/packages/components/src/button/style.scss#L305-L307

ecgan replied:
🤦‍♂️  I thought it was an issue with version 12.0.8 and not in the latest. Gotcha.

ecgan commented on js/src/components/contact-information/store-address-card.js:
```diff
@@ -0,0 +1,96 @@
+	let addressContent;
+	let description = '';
+
+	if ( ! isPreview ) {
+		description = __(
+			'Please confirm your store address for verification.',
+			'google-listings-and-ads'
+		);
+	}
```
💅  This is related to `isPreview`. In the PR description, you mentioned "the preview mode of StoreAddressCard will be used in the Setting page later". From your screen recording and screenshot there, my opinion is that the StoreAddessCard in Setup MC flow and in Settings page do look quite different, and I would create two separate components for them. With that approach, I wouldn't need to have `isPreview` props and the conditional checks related to it (which I find take a lil bit more time to understand and maintain), things would be more straightforward.

eason9487 replied:
Consider the deadline is already exceeded so far, and the implementation had been used and tested at the later PRs. I lean to keep it as is for now, and refactor it when developing the full version of contact information feature.

ecgan commented on js/src/components/contact-information/store-address-card.scss:
```diff
@@ -0,0 +1,12 @@
+.gla-store-address {
+	&__refresh-button {
+		position: absolute;
```
💅  Personally I'd only use absolute positioning when it's really needed. I would use CSS Flex and maybe edit the button's height / padding / margin to make the overall look nicer.

eason9487 replied:
The icon size is small. To make it has proper clickable area without raising up the container height and adjusting much more style to keep the container height. And there should be no more element at the right side of refresh button. I think it should be fine to set it `position: absolute` here.
![image](https://user-images.githubusercontent.com/17420811/127961982-9fccf8c4-2fda-495c-b481-f7c7813c070a.png)

ecgan reviewed (approved):
Overall okay, tested as per test instructions, work as expected. Left a few comments with ❔ 💅 .

One nitpick on the "Edit in Settings" button: the button padding is 6px, and the spacing between the text and the icon (the svg's margin-left) is 8px, which makes it look weird as there is a big gap in the middle. See screenshot below. Maybe we can address it in a separate issue.

![image](https://user-images.githubusercontent.com/417342/127909932-44fb1369-3a1a-49b8-905a-1a9ddd855fc3.png)

eason9487 committed:
Rename className for AppButton

Address https://github.com/woocommerce/google-listings-and-ads/pull/918#discussion_r681169915

tomalec commented on js/src/wcdl/subsection/title/index.scss:
```diff
@@ -1,4 +1,5 @@
 .wcdl-subsection-title {
+	position: relative;
```
FWIW, I double-checked the code of all our usages of `<Section.Title>`, no other seems to have non-text content, that could be broken by this rule.

tomalec commented on js/src/components/contact-information/store-address-card.scss:
```diff
@@ -0,0 +1,12 @@
+		position: absolute;
+		top: 0;
+		bottom: 0;
+		margin: auto;
```
:nail_care: :scroll: 
I think we could make the icon a (horizontal) center of focus area:

##### Suggested
![image](https://user-images.githubusercontent.com/17435/128050959-90f2f75a-c884-4e51-857f-90348c9d02a6.png)
##### Current
![image](https://user-images.githubusercontent.com/17435/128051055-10fa550a-5ff2-4ddf-9518-759ff00be34f.png)
```suggestion
		margin: auto;

		&.has-icon.has-text {
			justify-content: center;

			svg {
				margin: 0;
			}
		}
```

Or __preferably__, improve our `<AppButton>` implementation, to remove unnecessary whitespace text child nodes, if there are no children https://github.com/woocommerce/google-listings-and-ads/blob/trunk/js/src/components/app-button/index.js#L58-L61 so the `<button>` will not get `.has-text` class, therefore, the `left` alignment.

eason9487 replied:
Thanks! I took your suggested change first. It's more proper to improve the **AppButton** but there are a few tricky things that need much time to make sure all usages work well. I will open another PR later.

tomalec reviewed (approved):
Reviewed the code, tested locally. LGTM. Left [one cosmetic note](https://github.com/woocommerce/google-listings-and-ads/pull/918/files#r681924240) for something, that could be addressed later in a broader scope of `AppButton` implementation.

eason9487 committed:
Make the icon a horizontal center of focus area

Address https://github.com/woocommerce/google-listings-and-ads/pull/918#discussion_r681924240

Co-authored-by: Tomek Wytrębowicz <tomalecpub@gmail.com>
